### DISCUSSION TITLE
ESP32 == modbus tcp over wifi ==>> Huawei SUN2000 inverter

### DISCUSSION THREAD

**Discussion by mandgeo** [eModbus/eModbus#292] [General]
I've searched some terms, but haven't found something about this subject. Some of the discussions where related, but didn't found my answer.

What am I trying to do? I want to use an ESP32 to get data from a Huawei SUN2000 solar inverter. The connection is made over wifi. 
Why? Because my device is under warranty and I don't want to mess around with cables, even if it's just two wires and ground. Not that I'm afraid I'll do something wrong, but I don't want traces in logs. Do you have any idea if this device records such events?

What have I done so far? 
I've tested if a connection is possible. With my laptop, i've connected to the inverter wifi, launched QModMaster, set ip to 192.168.200.1, port to 6607, set ModBus mode to TCP, Unit ID (Slave Addr on older versions) to 0 requested data from address 30070... and I've got data.
With ESP32, I've adapted an tcp03async example, to ask data from address 30070 (2 words), but with no success. From my understanding, addRequest uses the server/master ID, but that is the id of the ESP. QModMaster works only if I set the correct slave address, unit 0. How can I point eModbus requests to unit 0? 
If I set the addRequest server ID to 0, it immediately returns E1 invalid server. Any other server ID returns, after some time, E0 request timeout.

What am I doing wrong? 

*** I think my level of understanding is average. I take bits and pieces, put them together, add something ... to satisfy my needs.

**Comment by Miq1** [comment]
Problem is, according to the Modbus standard server ID 0 is reserved for broadcast requests on Modbus RTU. Theoretically this will not matter on Modbus TCP, but the standard denies it for compatibility reasons. And so does eModbus...

Broadcast is possible on RTU with this library, but part of the Modbus regulations is, servers must not respond to a broadcast.

The only idea I had is to set up the request byte by byte, like:
```
ModbusMessage req;
req.add((uint8_t)0, READ_HOLD_REGISTER, (uint16_t)30070, (uint16_t)2);
Client.addRequest(req, (uint32_t)millis());
```
It may be this will fly below the checks the library normally does, but I can not predict if you successfully will receive any response. Give it a try, I would say and please keep me posted!

**Reply by mandgeo** [reply]
I'm getting data! First command always fails with timeout, but after that I get response for every request made at 500ms delay. Now I must find a way to convert that data :))

I think the trick with Huawei inverter is that the dongle acts as a repeater. For every request the dongle gets, it must know where to send that request, to the inverter, to the battery unit, to the smart meter, to another inverter in the system. *** just a theory based on limited knowledge.

**Reply by Miq1** [reply]
The dongle in fact is a Modbus TCP-to-RTU bridge, the inverter is on RTU. So the usual way was to assign a dedicated serverID to each device and route the requests accordingly. May be Huawei has deployed another non-standard way to request the devices present on the bus and to assign its own serverID scheme to those. This will relief field personnel from having to assign serverIDs on site.
But excellent to read you are getting data finally!
